### PR TITLE
add Type to kwargs internal argument

### LINKED ISSUE #39419: Lower keyword-body methods with tighter signature
Every method defined like `f(arg; kwargs...) = g(arg; kwargs...)` ends up having `kwargs` untyped in the body method, and so the conversion to `Symbol` is not inferrable. We've addressed this by adding a type-annotation to `merge(nt::NamedTuple, itr)`, but this is in a sense more fragile than it needs to be: it requires `itr` to have keys of specific types.

An alternative fix would be to change lowering. Specifically, in
```julia
julia> Meta.lower(Main, :(f(x; kwargs...) = g(x; kwargs...)))
:($(Expr(:thunk, CodeInfo(
    @ none within `top-level scope'
1 ─       $(Expr(:thunk, CodeInfo(
    @ none within `top-level scope'
1 ─     return $(Expr(:method, :f))
)))
│         $(Expr(:thunk, CodeInfo(
    @ none within `top-level scope'
1 ─     return $(Expr(:method, Symbol("#f#5")))
)))
│         $(Expr(:method, :f))
│         $(Expr(:method, Symbol("#f#5")))
│   %5  = Core.typeof(var"#f#5")
│   %6  = Core.Typeof(f)
│   %7  = Core.svec(%5, Core.Any, %6, Core.Any)
│   %8  = Core.svec()
│   %9  = Core.svec(%7, %8, $(QuoteNode(:(#= REPL[1]:1 =#))))
│         $(Expr(:method, Symbol("#f#5"), :(%9), CodeInfo(
```
is there any reason to have `%7` not be `Core.svec(%5, Base.Iterators.Pairs{Symbol}, %6, Core.Any)`?

### PR DESCRIPTION
**Pull request by vtjnash**
Fixes #39419

### CODE REVIEW

File: base/sysimg.jl [hunk 1]
```diff
@@ -23,8 +23,8 @@ let
     # Run with the `--exclude-jlls` option to filter out all JLL packages
     stdlibs = [
         # No dependencies
-        :ArgTools,
         :Artifacts,
+        :ArgTools,
```
Comment (Sacha0) [inline base/sysimg.jl:27]: Out of curiosity, why reorder? (You successfully nerd sniped me 😄.)

Comment (vtjnash) [reply base/sysimg.jl:27]: oops, that wasn't for this PR (I was timing ArgTools, and wondering why it was one of the slower packages to load)

**Comment by timholy** [comment]
Awesome, thanks so much! While you're at it, maybe change https://github.com/JuliaLang/julia/blob/ce5bd1c3dadcc7c733508a19a4b692b744d59a3d/base/namedtuple.jl#L293

to `k = Symbol(k)`?  While I think the typeassert was *practically* the right thing to do at the time, it's a weird limitation for an exported, generic function like `merge` and where `itr` can be anything. (Presumably we should support any keys from which a `Symbol` can be created, or at least those which can `convert` to a `Symbol`.)

If you're at all worried about the possibility of invalidation regressions, you can check the invalidations from PyPlot. I'm sure you've done this before many times, perhaps using more low-level tools, but in case it saves you any time here's a recipe of how I'd do this:
```julia
julia> using SnoopCompileCore

julia> invs = @snoopr using PyPlot;

julia> using SnoopCompile

julia> length(uinvalidated(invs))
1220
```
:scream: Something has already regressed! Looks like 8e6155150fc.

So the bar is low for this particular change, get it in before others raise the standards! :slightly_smiling_face:

File: base/essentials.jl [hunk 2]
```diff
@@ -23,6 +23,26 @@ An `AbstractDict{K, V}` should be an iterator of `Pair{K, V}`.
...
+Modifying the key-space of the underlying data may invalidate this object.
+"""
+struct Pairs{K, V, I, A} <: AbstractDict{K, V}
+    data::A
+    itr::I
+end
+Pairs{K, V}(data::A, itr::I) where {K, V, I, A} = Pairs{K, V, I, A}(data, itr)
+Pairs{K}(data::A, itr::I) where {K, I, A} = Pairs{K, eltype(A), I, A}(data, itr)
+Pairs(data::A, itr::I) where  {I, A} = Pairs{eltype(I), eltype(A), I, A}(data, itr)
+pairs(::Type{NamedTuple}) = Pairs{Symbol, V, NTuple{N, Symbol}, NamedTuple{names, T}} where {V, N, names, T<:NTuple{N, Any}}
```
Comment (JeffBezanson) [inline base/essentials.jl:39]: This seems a bit strange to have as a method; maybe just make it a constant instead?

File: src/julia-syntax.scm [hunk 3]
```diff
@@ -470,15 +470,16 @@
...
-          (mangled (let ((und (and name (undot-name name))))
-                     (symbol (string (if (and name (= (string.char (string name) 0) #\#))
-                                         ""
-                                         "#")
-                                     (or und '_) "#"
-                                     (string (current-julia-module-counter)))))))
+                  sparams))
+         (kw      (gensy))
+         (rkw     (if (null? restkw) (make-ssavalue) (symbol (string (car restkw) "..."))))
+         (restkw  (map (lambda (v) `(|::| ,v (call (top pairs) (core NamedTuple)))) restkw))
```
Comment (JeffBezanson) [inline src/julia-syntax.scm:476]: This could be merged with the first definition of restkw.

Comment (vtjnash) [reply src/julia-syntax.scm:476]: computing `rkw` needs the initial form (to `string`ify the variable name) before this wrapper is applied

**Comment by JeffBezanson** [comment]
> to `k = Symbol(k)`?

That would be a dramatic change, since we allow constructing a Symbol from anything by stringification.

**Comment by timholy** [comment]
Oh, funny I didn't know that. `convert(Symbol, k)` then?

**Comment by JeffBezanson** [comment]
Have you come across a use case for it? I'm surprised to see the Invalidation Czar himself propose this :)

**Comment by timholy** [comment]
Nah. For me Symbols live in their own world; until just now I also hadn't really that there are no objects in Base that convert to `Symbol`:
```julia
julia> methods(convert, (Type{Symbol}, Any))
# 1 method for generic function "convert":
[1] convert(::Type{T}, x::T) where T in Base at essentials.jl:171
```

All this explains why that change passed tests and hasn't caused other known problems. I was just motivated by the fact that it *looks* like a generic method, even if nothing takes advantage of that. If nothing else we could document that the keys must be `Symbol`.

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/39593#issuecomment-786422639) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/b97eb35_vs_b39d697/report.md). cc @christopher-dG

**Comment by timholy** [comment]
Thanks again, @vtjnash!

**Pull request merged**